src/V3Randomize.cpp

```diff
@@ -1323,23 +1323,77 @@ class RandomizeVisitor final : public VNVisitor {
                     if (!structSelp->dtypep()) structSelp->dtypep(smemberp->subDTypep());
                     randp = newRandStmtsp(fl, structSelp, nullptr);
                 }
-                if (stmtsp) {
-                    stmtsp->addNext(randp);
-                } else {
-                    stmtsp = randp;
-                }
+                stmtsp = stmtsp ? stmtsp->addNext(randp) : randp;
             }
             return stmtsp;
-        } else if (const auto* const unionDtp = VN_CAST(memberp ? memberp->subDTypep()->skipRefp()
-                                                                : exprp->dtypep()->skipRefp(),
-                                                        UnionDType)) {
+        } else if (const auto* const unionDtp = VN_CAST(memberDtp, UnionDType)) {
             if (!unionDtp->packed()) {
                 unionDtp->v3error("Unpacked unions shall not be declared as rand or randc."
                                   " (IEEE 1800-2023 18.4)");
                 return nullptr;
             }
             AstMemberDType* const firstMemberp = unionDtp->membersp();
             return newRandStmtsp(fl, exprp, nullptr, offset, firstMemberp);
+        } else if (AstDynArrayDType* const dynarrayDtp = VN_CAST(memberDtp, DynArrayDType)) {
+            V3UniqueNames* uniqueNamep = new V3UniqueNames{"__Vrandtemp"};
+            AstNodeDType* tempDTypep = dynarrayDtp;
+            AstCMethodHard* tempDynArrayMethodp = nullptr;
+            AstVar* randLoopIndxp = nullptr;
+            while (VN_CAST(tempDTypep, DynArrayDType)) {
+
+                AstVar* tempRandLoopIndxp
+                    = new AstVar{fl, VVarType::VAR, uniqueNamep->get(""),
+                                 dynarrayDtp->findBasicDType(VBasicDTypeKwd::UINT32)};
+
+                if (randLoopIndxp) {
+                    randLoopIndxp->addNext(tempRandLoopIndxp);
+                } else {
+                    randLoopIndxp = tempRandLoopIndxp;
+                }
+                tempDynArrayMethodp
+                    = tempDynArrayMethodp
+                          ? new AstCMethodHard{fl, tempDynArrayMethodp, "atWrite",
+                                               new AstVarRef{fl, tempRandLoopIndxp, VAccess::READ}}
+                          : new AstCMethodHard{
+                              fl, exprp, "atWrite",
+                              new AstVarRef{fl, tempRandLoopIndxp, VAccess::READ}};
+                tempDynArrayMethodp->dtypep(tempDTypep->subDTypep());
+                tempDTypep = tempDTypep->virtRefDTypep();
+            }
+            AstSelLoopVars* randLoopVarp = new AstSelLoopVars{
+                fl, new AstVarRef{fl, ((AstVarRef*)exprp)->varp(), VAccess::READ}, randLoopIndxp};
+            AstNodeStmt* stmtsp = new AstForeach{fl, randLoopVarp,
+                                                 newRandStmtsp(fl, tempDynArrayMethodp, nullptr)};
```

Comment:
To ensure this recurses currectly for sub-type like nested structs currently do, it may be worth to test arrays of structs and struct containing arrays, or some similar pattern.